conftest.py

```diff
@@ -248,3 +222,7 @@ def configure_testsetup(request):
         VIRT_DISPLAY.stop()
     else:
         yield
+
```

Comment:
the new approach is not to start on any test a mscolab service but once per test class. We need at the end to terminate all processes

Reply:
Is that really a good idea? We could run into issues where one test affects the outcome of another due to the shared mscolab instance, which could be a nightmare to debug. Also, I do not really like the implementation with this global variable here, couldn't this be a fixture sort of like this (pseudocode):
```
@pytest.fixture()
def mscolab_server():
    server = start_mscolab_server()
    yield server
    server.shutdown()
```
?

Every test that needs running mscolab server then should just use the fixture. That way we could even use the scope of the fixture to decide how "shared" we want the instance to be.

For what it's worth, I would prefer to get rid if test classes entirely, anyway. A yield fixture is much better suited for setup and teardown code, which leaves classes only as an organizational unit to group tests together. But modules, in turn, do that better.

Reply:
it is not a global variable, it is defined as module local variable. 
That's a great article about https://betterprogramming.pub/alternatives-to-using-globals-in-python-a3b2a7d5411b 

we need for xdist calls servers running on different ports, 

try it.

what do you mean by removing test classes entirely, I think setup and teardown may become complicated. But lets see.

Reply:
each test cleans up the database, and xdist is currently called on --dist loadfile 
The requirement by this is only different port per file. 

I also like to have different servers on each run, which I introduced before. 

Anyway any test needs independent from each other. So not having a test which creates data and another test which tries to vaildate something based on that.

Reply:
> what do you mean by removing test classes entirely, I think setup and teardown may become complicated. But lets see.

The recommended way (by pytest) to have setup and teardown code is using a yield fixture. You can read more about that here: https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended
They have an example there to demonstrate how user creation and deletion would work with fixtures, in an example testing a hypothetical email server. No class needed.

> Anyway any test needs independent from each other. So not having a test which creates data and another test which tries to vaildate something based on that.

Hypothetically, what would happen with these shared mscolab servers if we had a test for user deletion trying to delete user "UV10@uv10" in `Test_Mscolab_connect_window`, but also try to login to this user in `test_login`? My concern is that we might run into similar, but less obvious, race conditions in the future.

Reply:
xdist tests and reverse order by normal tests likly finds these dependencies. 

There were a lot in the past.

Reply:
> it is not a global variable, it is defined as module local variable.

In https://github.com/Open-MSS/MSS/blob/e60909c891f489a7ef1da309a63f85a547fa68bb/conftest.py#L226-L228 we terminate all processes once when we are finished with running the test suite. Isn't `MSCOLAB_PROCESSES` a global variable in the scope of the test session, then? We have one instance of this variable in which we accumulate all the running mscolab processes while executing the tests.

If I understand that right, with this PR we would also keep all the processes running until the end of the session, instead of stopping them as soon as they are not needed anymore.

Reply:
> > what do you mean by removing test classes entirely, I think setup and teardown may become complicated. But lets see.
> 
> The recommended way (by pytest) to have setup and teardown code is using a yield fixture. You can read more about that here: https://docs.pytest.org/en/6.2.x/fixture.html#yield-fixtures-recommended They have an example there to demonstrate how user creation and deletion would work with fixtures, in an example testing a hypothetical email server. No class needed.
> 
> > Anyway any test needs independent from each other. So not having a test which creates data and another test which tries to vaildate something based on that.
> 
> Hypothetically, what would happen with these shared mscolab servers if we had a test for user deletion trying to delete user "UV10@uv10" in `Test_Mscolab_connect_window`, but also try to login to this user in `test_login`? My concern is that we might run into similar, but less obvious, race conditions in the future.


connect_window and 
test_login has to setup the user,
after each test per teardown the database is cleared. 
Any test has to do all the steps for testing an approach.

when the tests are paralell running by xdist, any worker creates new tmp dirs

We had the described problems before, that's why tests currently not relay on some others.

Reply:
I would prefer more functional tests, without using the UI. At least everything what can be tested without the UI should be done.

Reply:
> At least everything what can be tested without the UI should be done.

I have to disagree on that. We should test at the "level" at which users interact with the application as well, and that is the UI. There is no point in testing if a login works if we don't test that the actual login button works also.

Or to phrase it differently: everything that can be done with the UI cannot be tested without it.

Reply:
we do this already in many ways also by the tutorials, But we also have a lot helper functions which also work without the UI and these should become tested if possibile without the UI.